mtl/mtl.py

```diff
@@ -0,0 +1,268 @@
+"""
+Copyright (c) 2023, Explainable Robotics Lab (ERL)
+ See license.txt file for license information.
+ @author: Gustavo A. Cardona, Cristian-Ioan Vasile
+"""
+
+import itertools as it
+
+import numpy as np
+from scipy.interpolate import interp1d
+from antlr4 import InputStream, CommonTokenStream
+
+from mtlLexer import mtlLexer
+from mtlParser import mtlParser
+from mtlVisitor import mtlVisitor
+
+
+class Operation(object):
+    '''STL operations'''
+    NOP, NOT, OR, AND, IMPLIES, UNTIL, EVENT, ALWAYS, PRED, BOOL = range(10)
+    opnames = [None, '!', '||', '&&', '=>', 'U', 'F', 'G', 'predicate', 'bool']
+    opcodes = {'!': NOT, '&&': AND, '||' : OR, '=>': IMPLIES,
+               'U': UNTIL, 'F': EVENT, 'G': ALWAYS}
+    opstrnames = [None, 'not', 'or', 'and', 'imply', 'until', 'event', 'always',
+                  'predicate', 'bool']
+    # negation closure of operations
+    negop = (NOP, NOP, AND, OR, AND, NOP, ALWAYS, EVENT, PRED, BOOL)
+
+    @classmethod
+    def getCode(cls, text):
+        ''' Gets the code corresponding to the string representation.'''
+        return cls.opcodes.get(text, cls.NOP)
+
+    @classmethod
+    def getString(cls, op):
+        '''Gets custom string representation for each operation.'''
+        return cls.opnames[op]
+
+    @classmethod
+    def getName(cls, op):
+        '''Gets custom long string name for each operation.'''
+        return cls.opstrnames[op]
+
+class MTLFormula(object):
+    '''Abstract Syntax Tree representation of an MTL formula'''
+
+    def __init__(self, operation, **kwargs):
+        '''Constructor'''
+        self.op = operation
+
+        if self.op == Operation.BOOL:
+            self.value = kwargs['value']
+        elif self.op == Operation.PRED:
+            self.variable = kwargs['variable']
+        elif self.op in (Operation.AND, Operation.OR):
+            self.children = kwargs['children']
+        elif self.op == Operation.IMPLIES:
+            self.left = kwargs['left']
+            self.right = kwargs['right']
+        elif self.op == Operation.NOT:
+            self.child = kwargs['child']
+        elif self.op in(Operation.ALWAYS, Operation.EVENT):
+            self.low = kwargs['low']
+            self.high = kwargs['high']
+            self.child = kwargs['child']
+        elif self.op == Operation.UNTIL:
+            self.low = kwargs['low']
+            self.high = kwargs['high']
+            self.left = kwargs['left']
+            self.right = kwargs['right']
+
+        self.__string = None
+        self.__hash = None
+
+    
+    # def negate(self):
+    #     '''Computes the negation of the MTL formula by propagating the negation
+    #     towards predicates.
+    #     '''
+    #     if self.op == Operation.BOOL:
+    #         self.value = not self.value
+    #     elif self.op == Operation.PRED:
```

Comment:
Here you would need to create a new MTLFormula object corresponding to a negation operation, and add the predicate as a child. The negation for all other operations remain the same.